File: build_plan.py
#!/usr/bin/env pypy3

import datetime
from datetime import datetime as dt
import csv
from collections import deque

stime = dt(2025, 9, 20)
etime = dt(2025, 12, 21)
latest_start_time = dt(2025, 11, 30)
min_layover = datetime.timedelta(minutes=50)
max_layover = datetime.timedelta(hours=18)
max_plan_dur = datetime.timedelta(days=45)
min_dest_airports = 12
max_dup_airports = 10
regional_arrival_cutoff = datetime.time(20, 0)
#home_arrival_cutoff = datetime.time(23, 30)
overnight_threshold = datetime.timedelta(hours=3)
overnight_check_time = datetime.time(3, 0)
min_trip_gap = datetime.timedelta(days=3)
max_trip_gap = datetime.timedelta(days=28)
allowed_overnight_airports = ['RDU', 'DCA', 'MCO', 'PVD', 'PIT']

start_airport = 'BOS'
regional_end = ['PVD', 'ORH']
end_airports = [start_airport] + regional_end

flight_graph = {}


class Flight:
    def __init__(self, src, dst, fnum, dtime, atime):
        self.src = src
        self.dst = dst
        self.fnum = fnum
        self.dtime = dtime
        self.atime = atime

    def __repr__(self):
        return f"Flight({self.src} -> {self.dst}, {self.fnum}, {self.dtime}, {self.atime})"


class ValidPlan:
    def __init__(self, flights, airports, total_dur, effective_dur):
        self.flights = flights
        self.airports = airports
        self.total_dur = total_dur
        self.total_days = calc_days_taken(flights)
        self.effective_dur = effective_dur

    def __repr__(self):
        return f"ValidPlan(flights={len(self.flights)}, airports={len(self.airports)}, total_duration={self.total_dur}, effective_duration={self.effective_dur})"

    def __str__(self):
        ret = ""
        ret += (f"Plan: {len(self.flights)} flights, {len(self.airports)} destinations:\n")
        for i, flight in enumerate(self.flights):
            if i < len(self.flights) - 1:  # Not the last flight
                next_flight = self.flights[i + 1]
                layover = next_flight.dtime - flight.atime
                ret += (f"  {flight.src} -> {flight.dst} (B6 {flight.fnum:04d}) "
                        f"{flight.dtime.strftime('%m.%d %H:%M')} -> "
                        f"{flight.atime.strftime('%m.%d %H:%M')} "
                        f"[layover: {layover}]\n")
            else:  # Last flight
                ret += (f"  {flight.src} -> {flight.dst} (B6 {flight.fnum:04d}) "
                        f"{flight.dtime.strftime('%m.%d %H:%M')} -> "
                        f"{flight.atime.strftime('%m.%d %H:%M')}\n")
        ret += (f"  Total duration: {self.total_dur}\n"
                f"  Effective duration: {self.effective_dur}\n"
                f"  Days taken: {self.total_days}\n"
                f"  Airports visited: {sorted(self.airports)}\n")
        return ret


def build_flight_graph(csv_file):
    # CSV format:
    # departure_airport,arrival_airport,flight_number,departure_time,arrival_time
    flight_graph = {}
    with open(csv_file, 'r') as f:
        reader = csv.reader(f, delimiter=',')
        next(reader)  # Skip header row
        for row in reader:
            src, dst, fnum, dept_timestr, arr_timestr = row
            dtime = dt.strptime(dept_timestr, '%Y-%m-%d %H:%M:%S')
            atime = dt.strptime(arr_timestr, '%Y-%m-%d %H:%M:%S')

            if dtime < stime or atime > etime:
                continue

            flight = Flight(src, dst, int(fnum), dtime, atime)
            if src not in flight_graph:
                flight_graph[src] = []
            flight_graph[src].append(flight)
    return flight_graph


def get_valid_outgoing(incoming, graph):
    ret = []
    cur = incoming.dst
    for outgoing in graph[cur]:
        layover = outgoing.dtime - incoming.atime
        if min_layover >= layover or layover >= max_layover:
            continue

        # Check if overnight and if airport is allowed
        # overnight meaning over overnight check threshold and layover period contains overnight_check_time
        if layover > overnight_threshold:
            overnight_check_datetime = dt.combine(
                outgoing.atime.date(), overnight_check_time)
            if incoming.atime <= overnight_check_datetime <= outgoing.dtime:
                if cur not in allowed_overnight_airports:
                    continue
        ret.append(outgoing)
    return ret


def calc_days_taken(flights):
    days = {flight.dtime.date() for flight in flights}
    days |= {flight.atime.date() for flight in flights}
    return len(days)


def calc_effective_duration(flights):
    # count total duration except start airport layovers exceeding min_trip_gap
    dur = datetime.timedelta(0)
    for i, flight in enumerate(flights):
        dur += flight.atime - flight.dtime
        if i < len(flights) - 1:
            next_flight = flights[i + 1]
            layover_duration = next_flight.dtime - flight.atime
            # Only add layover if not at start airport
            if flight.dst != start_airport or layover_duration <= min_trip_gap:
                dur += layover_duration
    return dur


def get_new_bos_outgoing(intime, graph):
    ret = []
    for outgoing in graph[start_airport]:
        gap = outgoing.dtime - intime
        if min_trip_gap <= gap <= max_trip_gap:
            ret.append(outgoing)
    return ret


def is_valid_endpoint(flight):
    if flight.dst not in end_airports:
        return False

    if flight.atime > etime or flight.dtime < stime:
        return False

    if flight.dst in regional_end:
        # For regional airports, check if arrival time is before cutoff
        if flight.atime.time() > regional_arrival_cutoff:
            return False

    return True


def is_better_plan(old_plan: ValidPlan, new_plan: ValidPlan):
    # compare in order: # of days, # of flights, effective duration
    if new_plan.total_days < old_plan.total_days:
        return True
    if new_plan.total_days == old_plan.total_days:
        if len(new_plan.flights) < len(old_plan.flights):
            return True
        if len(new_plan.flights) == len(old_plan.flights):
            if new_plan.effective_dur < old_plan.effective_dur:
                return True
    return False


def trip_numdays(trip):
    if not trip:
        return 0
    start_date = trip[0].dtime.date()
    end_date = trip[-1].atime.date()
    return (end_date - start_date).days + 1


def calc_trip_dur(trip):
    return trip[-1].atime - trip[0].dtime


def main():
    graph = build_flight_graph('flights.csv')
    q = deque()

    # Initialize with flights from start airport
    for flight in graph[start_airport]:
        if latest_start_time.date() >= flight.dtime.date() >= stime.date():
            q.appendleft(([flight], {flight.dst}))

    best_plan = None
    i = 0
    while q:
        i += 1
        if i % 10_000_000 == 0:
            print(f".", end='', flush=True)
        if i % 1_000_000_000 == 0:
            print(f"={i//1_000_000_000}B done; {len(q)} queue")

        cur_plan, visited_airports = q.pop()

        # Check if current trip is a valid endpoint
        if is_valid_endpoint(cur_plan[-1]):
            # are we done yet
            if len(visited_airports) >= min_dest_airports:
                new_plan = ValidPlan(
                    flights=cur_plan.copy(),
                    airports=visited_airports,
                    total_dur=calc_trip_dur(cur_plan),
                    effective_dur=calc_effective_duration(cur_plan)
                )
                if not best_plan or is_better_plan(best_plan, new_plan):
                    best_plan = new_plan
                    print(f"\nBest plan so far: {best_plan}")
            # we're not done, start new trip
            else:
                for flight in get_new_bos_outgoing(cur_plan[-1].atime, graph):
                    new_visited = visited_airports | {flight.dst}
                    q.append((cur_plan + [flight], new_visited))
                pass

        # Continue searching if we haven't reached max duplicates
        if len(cur_plan) < len(visited_airports) + max_dup_airports:
            # Check total trip duration
            trip_duration = cur_plan[-1].atime - cur_plan[0].dtime
            if trip_duration <= max_plan_dur:
                # Get valid outgoing flights
                if cur_plan[-1].dst in graph:
                    for next_flight in reversed(get_valid_outgoing(cur_plan[-1], graph)):
                        new_visited = visited_airports | {next_flight.dst}
                        q.append((cur_plan + [next_flight], new_visited))

    print(f"\nSearch complete. Best plan found:\n{best_plan}")


if __name__ == "__main__":
    main()

File: build_plan.cpp
#include "absl/container/flat_hash_map.h"
#include "absl/flags/flag.h"
#include "absl/flags/parse.h"
#include "absl/time/civil_time.h"
#include "absl/time/time.h"
#include "fmt/core.h"
#include "rapidcsv.h"
#include <_abort.h>
#include <algorithm>
#include <csignal>
#include <ctime>
#include <deque>
#include <string>
#include <thread>
#include <unordered_map>
#include <vector>

using u8 = uint8_t;
using u16 = uint16_t;
using u32 = uint32_t;
using u64 = uint64_t;
using tp = absl::Time;
using str = std::string;
template <typename T> using vec = std::vector<T>;
template <typename K, typename V> using umap = std::unordered_map<K, V>;
template <typename T> using sptr = std::shared_ptr<T>;

const u32 DEST_CAP = 25;
const u32 MIN_DESTS = 15;
const u32 MAX_DUP_DESTS = 3;
const auto MAX_TRIP_DURATION = absl::Hours(7 * 24);

// Configuration constants
const auto START_TIME =
    absl::FromCivil(absl::CivilHour(2025, 9, 16), absl::UTCTimeZone());
const auto END_TIME =
    absl::FromCivil(absl::CivilHour(2025, 12, 15), absl::UTCTimeZone());
const auto MIN_DAY_LAYOVER = absl::Minutes(50);
const auto MAX_DAY_LAYOVER = absl::Hours(18);
const auto MIN_NIGHT_LAYOVER = absl::Hours(8);
const auto MAX_NIGHT_LAYOVER = absl::Hours(18);
const auto MIN_HOME_LAYOVER = absl::Hours(3 * 24 + 6);
const auto MAX_HOME_LAYOVER = absl::Hours(7 * 24);

// const u32 REGIONAL_ARRIVAL_CUTOFF_HOUR = 20;
const auto OVERNIGHT_THRESHOLD = absl::Hours(3);
const u32 OVERNIGHT_CHECK_HOUR = 3;

const str START_AIRPORT = "BOS";
const vec<str> REGIONAL_END = {"PVD", "ORH"};
const vec<str> ALLOWED_OVERNIGHT_AIRPORTS = {"RDU", "DCA", "BUF", "PVD", "PIT"};

auto timezone_lookup(str city) -> absl::TimeZone
{
    absl::TimeZone tz;
    if (city == "ABQ" || city == "DEN")
        absl::LoadTimeZone("America/Phoenix", &tz);
    else if (city == "BQN" || city == "PSE" || city == "SJU" || city == "STT" ||
             city == "STX")
        absl::LoadTimeZone("America/Puerto_Rico", &tz);
    else if (city == "DFW" || city == "DTW" || city == "IAH" || city == "MSY" ||
             city == "MKE" || city == "ORD")
        absl::LoadTimeZone("America/Chicago", &tz);
    else if (city == "LAS" || city == "PHX" || city == "LAX")
        absl::LoadTimeZone("America/Los_Angeles", &tz);
    else
        absl::LoadTimeZone("America/New_York", &tz);
    return tz;
}

struct Flight {
    str src;
    str dst;
    u16 fnum;
    absl::Duration dur;

    tp dep_time;
    tp arr_time;
    absl::CivilDay dep_day;
    absl::CivilDay arr_day;
    absl::TimeZone dep_tz;
    absl::TimeZone arr_tz;

    auto to_string() const -> str
    {
        return fmt::format("{:>4}: {} -> {}, {} -> {}, {}", fnum, src, dst,
                           absl::FormatTime("%m.%d %H:%M", dep_time, dep_tz),
                           absl::FormatTime("%m.%d %H:%M", arr_time, arr_tz),
                           absl::FormatDuration(dur));
    }

    auto is_valid_end() const -> bool
    {
        if (dst != START_AIRPORT || !absl::c_linear_search(REGIONAL_END, dst))
            return false;
        // TODO check regional arrival cutoff time
        return true;
    }

    auto is_overnight_to(Flight &oth) -> bool
    {
        assert(this->dst == oth.src);
        auto layover = oth.dep_time - this->arr_time;
        if (layover < OVERNIGHT_THRESHOLD)
            return false;

        // check if the layover is overnight at the shared airport
        auto cday = this->arr_day;
        auto compare_point = absl::FromCivil(
            absl::CivilHour(cday.year(), cday.month(), cday.day(),
                            OVERNIGHT_CHECK_HOUR, 0),
            this->arr_tz);

        return this->arr_time <= compare_point && oth.dep_time >= compare_point;
    }
};

class Plan : public std::enable_shared_from_this<Plan>
{
  public:
    sptr<Plan> prev;     // previous plan in the chain
    Flight *last_flight; // last flight in the plan
    u8 num_flights;      // total number of flights
    u8 unique_dest;      // unique airports
    u8 num_days;         // whole days used
    absl::Duration dur;  // total trip duration

    static auto new_root(Flight *f)
    {
        auto ret = std::make_shared<Plan>();
        ret->prev = nullptr;
        ret->last_flight = f;
        ret->num_flights = 1;
        ret->unique_dest = 1;
        ret->num_days = 1;
        ret->dur = f->dur;
        return ret;
    }

    auto new_append(Flight *nxt)
    {
        auto ret = std::make_shared<Plan>();
        ret->prev = shared_from_this();
        ret->last_flight = nxt;
        ret->num_flights = (u8)(this->num_flights + 1);
        ret->unique_dest = (u8)(this->unique_dest + 1);
        ret->num_days = this->num_days;
        ret->dur = this->dur + nxt->dur;

        // search through all previous flights to see if this airport is new
        // and adjust unique_dest accordingly
        // for (auto &f : this->flights) {
        //     if (f->dst != nxt->dst)
        //         continue;
        //     ret.unique_dest--;
        //     break;
        // }

        Plan *p = this;
        while ((p = p->prev.get()) != nullptr) {
            if (p->last_flight->dst != nxt->dst)
                continue;
            ret->unique_dest--;
            break;
        }

        // check if we hit a new day
        if (nxt->dep_day != this->last_flight->arr_day)
            ret->num_days++;
        if (nxt->arr_day != nxt->dep_day)
            ret->num_days++;

        // adjust duration for layover if it's not the home city
        if (nxt->src != START_AIRPORT) {
            auto layover = nxt->dep_time - this->last_flight->arr_time;
            ret->dur += layover;
        }
        return ret;
    }

    auto is_better(const Plan &other) const -> bool
    {
        // compare in order:
        // - most unique destinations
        // - fewest days taken
        // - fewest flights + overnight layovers (TODO)
        // - shortest effective duration
        if (this->unique_dest != other.unique_dest)
            return this->unique_dest > other.unique_dest;
        if (this->num_days != other.num_days)
            return this->num_days < other.num_days;
        if (this->num_flights != other.num_flights)
            return this->num_flights < other.num_flights;
        return this->dur < other.dur;
    }

    auto to_string() const -> str
    {
        vec<Flight *> flight_list;
        for (auto p = this; p != nullptr; p = p->prev.get())
            flight_list.push_back(p->last_flight);
        std::reverse(flight_list.begin(), flight_list.end());

        str flight_str;
        for (u32 i = 0; i < flight_list.size() - 1; ++i) {
            flight_str += flight_list[i]->to_string();
            auto layover =
                flight_list[i + 1]->dep_time - flight_list[i]->arr_time;
            flight_str +=
                fmt::format(" [layover: {}]\n", absl::FormatDuration(layover));
        }
        flight_str += flight_list.back()->to_string() + "\n";

        return fmt::format("Plan: {} flights, {} destinations, {} days taken, "
                           "{} effective duration\n{}",
                           num_flights, unique_dest, num_days,
                           absl::FormatDuration(dur), flight_str);
    }
};

using CityGraph = umap<str, vec<Flight>>;
using FlightGraph = umap<Flight *, vec<Flight *>>;
const str TIME_FORMAT = "%Y-%m-%d %H:%M:%S";

auto build_city_graph(const str &filename) -> CityGraph
{
    rapidcsv::Document doc(filename);
    CityGraph city_graph;

    for (u32 i = 0; i < doc.GetRowCount(); ++i) {
        Flight f;
        f.src = doc.GetCell<str>("departure_airport", i);
        f.dst = doc.GetCell<str>("arrival_airport", i);
        f.fnum = (u16)doc.GetCell<u32>("flight_number", i);

        auto depart_tz = timezone_lookup(f.src);
        auto arrive_tz = timezone_lookup(f.dst);
        absl::ParseTime(TIME_FORMAT, doc.GetCell<str>("departure_time", i),
                        depart_tz, &f.dep_time, nullptr);
        absl::ParseTime(TIME_FORMAT, doc.GetCell<str>("arrival_time", i),
                        arrive_tz, &f.arr_time, nullptr);
        f.dur = f.arr_time - f.dep_time;

        f.dep_day = absl::ToCivilDay(f.dep_time, depart_tz);
        f.arr_day = absl::ToCivilDay(f.arr_time, arrive_tz);

        if (!city_graph.contains(f.src))
            city_graph[f.src] = vec<Flight>();
        if (f.dep_time > START_TIME && f.arr_time < END_TIME)
            city_graph[f.src].push_back(f);
    }

    return city_graph;
}

auto get_valid_outgoing(Flight &inc, CityGraph &graph) -> vec<Flight *>
{
    vec<Flight *> ret;
    for (auto &f : graph[inc.dst]) {
        // skip flights that depart before the arrival of the incoming flight
        if (f.dep_time < inc.arr_time)
            continue;

        // skip flights with too short or too long layovers
        auto layover = f.dep_time - inc.arr_time;
        if (inc.is_overnight_to(f) &&
            (layover < MIN_NIGHT_LAYOVER || layover > MAX_NIGHT_LAYOVER ||
             !absl::c_linear_search(ALLOWED_OVERNIGHT_AIRPORTS, f.src)))
            continue;
        else if (layover < MIN_DAY_LAYOVER || layover > MAX_DAY_LAYOVER)
            continue;

        ret.push_back(&f);
    }
    std::sort(ret.begin(), ret.end(),
              [](Flight *a, Flight *b) { return a->dep_time > b->dep_time; });
    return ret;
}

auto get_outgoing_new_trip(Flight &inc, CityGraph &g) -> vec<Flight *>
{
    assert(inc.dst == START_AIRPORT);

    // get all flights from the start airport that depart after START_TIME
    vec<Flight *> ret;
    for (auto &f : g[START_AIRPORT]) {
        if (f.dep_time >= inc.dep_time + MIN_HOME_LAYOVER &&
            f.dep_time <= inc.dep_time + MAX_HOME_LAYOVER)
            ret.push_back(&f);
    }
    std::sort(ret.begin(), ret.end(),
              [](Flight *a, Flight *b) { return a->dep_time < b->dep_time; });
    return ret;
}

auto build_flight_graph(CityGraph &city_graph) -> FlightGraph
{
    FlightGraph ret;
    for (auto &[_, flights] : city_graph)
        for (auto &f : flights)
            ret[&f] = get_valid_outgoing(f, city_graph);

    return ret;
}

auto search(u32 worker_num, CityGraph &cg, FlightGraph &fg,
            vec<Flight *> &start_q) -> Plan
{
    Plan best;
    std::deque<sptr<Plan>> q;
    for (auto start_flight : start_q)
        q.push_front(Plan::new_root(start_flight));

    u64 i = 0;
    while (!q.empty()) {
        // report progress
        i++;
        if (i % 10'000'000 == 0) {
            fmt::print(".");
            fflush(stdout);
        }
        if (i % 1'000'000'000 == 0)
            fmt::print("Worker {}: processed {}B, {} in queue\n", worker_num,
                       i / 1'000'000'000, q.size());

        auto cur = q.back();
        q.pop_back();

        // abandon path if it violates constraints
        if (cur->dur > MAX_TRIP_DURATION ||
            cur->num_flights > cur->unique_dest + MAX_DUP_DESTS)
            continue;

        // check if we're done with a trip, either completely or if we
        // should start a new one
        if (cur->last_flight->is_valid_end()) {
            if (cur->unique_dest >= MIN_DESTS && cur->is_better(best)) {
                best = *cur;
                fmt::print("New best plan found:\n{}\n", best.to_string());
            }

            // should we start a new trip?
            if (cur->unique_dest < DEST_CAP)
                for (auto &nxt : get_outgoing_new_trip(*cur->last_flight, cg))
                    q.push_back(cur->new_append(nxt));
        }

        // continue searching
        for (auto nxt : fg[cur->last_flight]) {
            // do some manual pruning

            // 1. no turn-arounds in the first half of the journey
            if (cur->num_flights < DEST_CAP / 2 &&
                nxt->dst == cur->last_flight->src)
                continue;

            // 2. no going back to start airport until 2nd half
            if (cur->num_flights < DEST_CAP / 2 && nxt->src == START_AIRPORT)
                continue;

            // append the next flight to the current plan
            q.push_back(cur->new_append(nxt));
        }
    }

    return best;
}

ABSL_FLAG(u32, num_workers, 0,
          "Number of workers to use for parallel search. 0 to use all cores.");

int main(int argc, char **argv)
{
    absl::ParseCommandLine(argc, argv);
    fmt::print("Current working directory: {}\n",
               std::filesystem::current_path().string());

    fmt::print("Building city graph...\n");
    auto city_graph = build_city_graph("flights.csv");
    fmt::print("Building flight graph...\n");
    auto flight_graph = build_flight_graph(city_graph);

    fmt::print("Starting search...\n");

    auto workers = absl::GetFlag(FLAGS_num_workers);
    if (workers == 0)
        workers = std::thread::hardware_concurrency() - 1;

    vec<Flight *> starting_flights;
    for (auto &f : city_graph[START_AIRPORT])
        if (f.dep_time >= START_TIME && f.dep_time <= END_TIME)
            starting_flights.push_back(&f);

    auto num_each = starting_flights.size() / workers;
    fmt::println("Starting {} workers, each with {} flights...", workers,
                 num_each);

    vec<std::jthread> threads;
    if (workers == 1) {
        // if only one worker, just run the search directly
        auto best = search(0, city_graph, flight_graph, starting_flights);
        fmt::print("Best plan found:\n{}\n", best.to_string());
        return 0;
    }

    for (u32 i = 0; i < workers; ++i) {
        threads.emplace_back([&, i]() {
            vec<Flight *> chunk;
            auto start_idx = i * num_each;
            auto end_idx =
                std::min(start_idx + num_each, starting_flights.size());
            for (auto j = start_idx; j < end_idx; ++j)
                chunk.push_back(starting_flights[j]);

            auto best = search(i, city_graph, flight_graph, chunk);
            fmt::print("Worker {} finished with best plan:\n{}\n", i,
                       best.to_string());
        });
    }

    for (auto &t : threads)
        t.join();

    return 0;
}
